src/Microsoft.Data.SqlClient/src/Microsoft/Data/ProviderBase/DbMetaDataFactory.cs

```diff
@@ -175,13 +177,18 @@ private DataTable ExecuteCommand(DataRow requestedCollectionRow, string[] restri
                     Locale = CultureInfo.InvariantCulture
                 };
 
+                cancellationToken.ThrowIfCancellationRequested();
+#if NET
+                DataTable schemaTable = isAsync ? await reader.GetSchemaTableAsync(cancellationToken).ConfigureAwait(false) : reader.GetSchemaTable();
+#else
```

Comment:
Would it make sense to have the .NET Framework path also use an async API here? Unless I'm missing something, the exact same methods/code could be define for both, with the sole extension that in .NET Framework it wouldn't override?

Reply:
I can do that; I thought it might look a little clumsy for the .NET Framework version of Microsoft.Data.SqlClient to have a dedicated async method which isn't present on System.Data.SqlClient and which doesn't actually run asynchronously.

(GetSchemaTable just constructs a DataTable from the cached metadata - there's nothing which we can make async.)

Reply:
Running this through CI now...

The IDbColumnSchemaGenerator method will need to be done after (or possibly as part of) #2967.

Reply:
> GetSchemaTable just constructs a DataTable from the cached metadata - there's nothing which we can make async

Oh I must be misunderstanding what this is about then - I thought the idea here is to actually get the schema information (asynchronously).

In any case, the idea is to have the least amount of separate code paths here for .NET Framework vs. modern .NET...

Reply:
I implemented this, but on re-checking, we access a property within SqlDataReader which performs a read from the network. I'm going to revert the SqlDataReader changes tomorrow so that I don't introduce async-over-sync within the new API. I'll submit the new SqlDataReader API in a subsequent PR, which will need more review since it'll involve touching the state machine.

SqlConnection.GetSchemaAsync is much simpler, since it just layers atop SqlCommand. This PR can still introduce that safely.